### ISSUE TITLE
FR: Add "Search in" filter capability

### ISSUE THREAD

**Issue by joeytakeda** [projectEndings/staticSearch#20]
MH and JT discussed this in person today and thought that projects may want the ability to "search in" particular fields--document titles, just the footnotes, metadata, et cetera. This is slightly different from the `text` filter that's next on the TODO list: while the text filter provides a non-stemmed searching ability for metadata (i.e. search for names, et cetera), the "Search in" would give users the ability to perform a standard word search in a particular field.

How we think we'd implement this:

* Add a step in the config file where users could specify, via Xpath, the various search in filters. Something maybe like this:
```
<searchIn>
<rule xpath="//div[contains-token(@class,'footnote')]" label="Footnotes"/>
<rule xpath="//div[id='metadata']" label="Metadata"/>
</searchIn>
```
* The XSLT would then add those as a `@data-staticSearch-searchIn` attribute in the tokenization step.

* When we create the KWICs, we would then look to see if the word token inherits any searchIn filters (i.e. `$span[ancestor::*[@data-staticSearch-searchIn]]`). If so, we'd add that as an array for the KWIC

* The search page creation would then have to determine some way to create these checkbox fields; they might just be keyed to the label values

* If a search in has been checked in the JS, then we see if the returned KWIC has the checked searched in.

Note that is not a necessary feature, but it is on the docket as an enhancement once all the necessary search features have been implemented.

**Comment by martindholmes** [comment]
Commit #90b3028 adds `@label` to the `<context>` element. Nothing is done with it yet in processing.

**Comment by martindholmes** [comment]
I think this functionality can be achieved entirely using contexts. If you're wanting to search within a component of a page, then it must perforce be a context (kwics shouldn't overflow it), so you're already defining it as a context; all you then need to do is give it a label, and we could then create a search option from that label.

**Comment by martindholmes** [comment]
@joeytakeda Switched this from the 1.1 milestone to the 1.2 milestone; we have the basics in place, but we need a more detailed plan, along with additions to the test suite, so it'll be a while, and none of our projects are calling for it right now. Switch it back if you disagree.

**Comment by martindholmes** [comment]
@joeytakeda I'm going to switch this to Blue Sky, since we have no projects needing it, and there are other approaches that are actually more flexible, based on multiple search pages, as implemented in DVPP.

**Comment by martindholmes** [comment]
Another +1 for this feature, arising out of the TEI Guidelines search: being able to search only examples whatever their context would be useful, as well as being able to search in containing/contained by contexts.

**Comment by joeytakeda** [comment]
Another +1 for this feature is for the Lyon in Mourning project, which has different kinds of annotations/abstracts for an item in the collection and would like to be able to the search just within a particular kind of note on a page.

**Comment by martindholmes** [comment]
Will the multiple-search-page approach not work for Lyon in Mourning?

**Comment by joeytakeda** [comment]
A few questions about this:

* Should context labels be inherited? E.g. say you have two contexts--footnotes and abstracts. Would a search within "abstract" return results in a footnote in an abstract? (I would hesitantly say yes)

* Can you search within multiple contexts? E.g. can I search for a term in either footnotes or abstracts?

* Do "Search within" searches always require an actual search query, or can you do a document only search? On one hand, it makes no sense to "search within" and then not search for anything (i.e. of course there should be no results); on the other hand, I imagine that a queryless "search within" could yield all of the documents that contain that context, but that kind of functionality is already handled mostly by boolean filters anyway. So I'd say "No--you need a search query for a search within to work".

* And what would a search within look like in the controls? I can imagine two possibilities, which depend a bit on whether we allow for multiple contexts to be searched at a time. the first would be to use a checkbox system like an ssDesc, but that we'd probably want it to be visually distinct in some way so that it's clear you're not searching for "Abstract" documents, but within an abstract of a document (i.e. maybe have the "Search within" check boxes stretch the entire width of the query box rather and have a border of some sort). The other option is a having a dropdown select list beside the query, which would be less obtrusive but possibly confusing (since you're searching within a document, rather than searching across a document type, which could be the expectation)

**Comment by joeytakeda** [comment]
> Will the multiple-search-page approach not work for Lyon in Mourning?

It would be a bit overkill: there are editorial abstracts about the item as well as different kinds of quotations /exceptions taken from the text (i.e. a quotation that an encoder has flagged as about media; a quotation they've flagged as expressing some sentimentality; etc). We could have multiple search pages, but we'd end up having about 5 different search engines, all of which only differ from the context that they're searching (i.e. they're all searching across the document collection, but looking within a particular phrase).

Another rationale is to be able to search by title of document (especially since the the items themselves have up to three titles: the ones written on the document itself, the title as indexed by Forbes, and the 19th century title). 
Searching for something like "MacDonald" would be difficult to sift through, since almost every document mentions a MacDonald (and in many cases, mentions many MacDonalds multiple times) but far fewer have MacDonald in the title (i.e. letters sent to or from a MacDonald). if you could search just within titles, then it'd make finding a single item in the collection much easier. Of course, that could also be an additional search page, but it feels like a bit of overkill.

**Comment by martindholmes** [comment]
Re inheritance of contexts: This is really hard to figure out. If you choose to search within abstracts but not within document titles (assuming those are both context options), should we exclude a result that's in an abstract because it happens to be in a document title at the same time? Ditto for the reverse case? I think the only workable answer is that you return results from the context even if they also happen to be in an unselected context. Better to retrieve more than less.

Re multiple contexts: I would say of course. Any KWIC must carry the entire hierarchy of pre-defined contexts in which it appears, and if any of those contexts is checked, the item is a hit.

I agree that "search within" requires a search query string. If you want to allow people to find "only documents with abstracts", that's an ssDesc filter.

I think the GUI should look just like ssDesc, but it's up to the user to label their contexts clearly enough so that it's obvious you're not searching *for documents with abstracts* but *within abstracts in documents*.

**Comment by joeytakeda** [comment]
I agree with all of @martindholmes' points above. So just to work through an example:

Assume the HTML is like so:

```xml
<main>
    <div id="abstract">
         <h2 id="heading">Summary of Article</h2>
         <p>This article is about searching.</p>
     </div>
     <div id="content">
           <p>This article concerns searching.</p>
      </div>
</main>
```

And contexts defined in the KWIC like so:

```
<context match="div[@id='abstract']" label="Abstracts"/>
<context match="h2[@id='heading']" label="Section headings"/>
```

A search for `article` with no contexts selected would return 3 hits:

* Summary of **Article**
* This **article** is about searching
* This **article** concerns searching

A search for `article` with "Abstracts" selected would return 2 hits:

* Summary of **Article**
* This **article** is about searching

A search for `article` with "Section headings" selected would return a single hit:

* Summary of **article**

And a search for `article` with "Section headings" and "Abstracts" selected (which would mean, like the ssDesc, any contexts that are in "Section headings" OR "Abstracts") would return 2 hits (the same as the search for `article` just with "Abstracts" selected)

* Summary of **Article**
* This **article** is about searching

In terms of processing, we run into the classic problem of ids for those. I think we'd want to take a slightly different approach to the id creation since we already know all of the contexts available (i.e. we've specified contexts in the config rather than in the HTML itself). I think what we do is:

1. During the config creation step (where we create the config XSL for that instance of static search), we create a static map of ids to contexts:

```xsl
<xsl:variable name="ssContextsMap" select="map{
'ssCtx1': "Document headings",
'ssCtx2': "Abstracts"
}"/>
```

2. When the document is tokenized, each context block is given an id (I'm using the word "in" since the JSON already uses "contexts" for all sorts of things)

```
<div id="abstract" data-ssin="ssCtx1">
<h2 id="heading" data-ssin="ssCtx2">Abstract</h2>
<p>This article is about searching.</p>
</div>
```

3. When we create the token files, we then add all of the ids that the hit is in:

```
{
 "form" : "Article",
 "weight" : "1",
  "pos" : 1102,
  "in": ['ssCtx2', 'ssCtx1']
  "context" : "This <mark>article</mark> is about searching",
  "fid" : "abstract"
 }
```

And when we create the search page, we use the map we created in the config step to make the HTML checkboxes (iterating through the keys, and creating a list item for each key and with its value as the text for the checkbox, sorting by the values).

I think in the JS, if any of the "Search in" filters are checked, we'd:
* Discard any hits that don't have an "in" key
* For each hit, check to see if the hit's "in" array intersects with the selected "Search in" array (using the XSet) 

Thoughts?

**Comment by martindholmes** [comment]
This looks exactly right to me. One thing that's become slightly clearer in thinking again about it is that this feature is really "Search ONLY in," and of course that selecting multiple contexts ORs them rather than ANDing them.

**Comment by joeytakeda** [comment]
Good point—we should make the default label "Search only in" to make that clear (as well as outline that in the documentation).

I've just started a new branch for this (iss20-search-only-in) and will start plugging away at it.

**Comment by joeytakeda** [comment]
Had a clear afternoon, so I've been able to make good progress on this. I think I now have a working implementation in the branch, but it still needs some test cases and documentation.

**Comment by martindholmes** [comment]
Nice work. In the normalization branch, I'm stuck on three annoying errors in the French stemmer. Hoping to get a run at that some time this week, then I'll be able to look at this stuff.

**Comment by martindholmes** [comment]
@joeytakeda I've added two contexts (abstracts and minutes) to the Colonial Despatches build and set it to use the issue 20 branch, so we can see a test with a large document set. Should be done in two or three hours from now.

**Comment by martindholmes** [comment]
It looks good and it seems to be working great. Time for a pull request?

**Comment by joeytakeda** [comment]
Excellent--I'll add a couple test cases and some documentation, then do a PR for the whole thing.

**Comment by martindholmes** [comment]
Might be worth doing a pull from dev to see if there are any conflicts with the recently-merged NFC stuff before the PR.

**Comment by joeytakeda** [comment]
Merged the NFC stuff into dev, added some documentation and tests (which made me realize the stuff was a bit broken still--scores weren't being updated and phrasals [and wildcards] weren't properly handled), and have now added a PR.
